### PR TITLE
P2: Fixes the conflict between Flash API and XIP

### PR DESCRIPTION
**Pull request by YutingYou**
### Problem

When the user app is writing or erasing the Flash with the Flash API, in the meanwhile,  the system thread is trying to read a lot of data via XIP e.g. CRC32 calculation, we'll see a crash on P2.

### Solution

Since our firmware is running on the PSRAM, XIP will not be used most of the time on KM4, any accesses to the external flash will use the Flash API. So we were going to disable XIP altogether, however, given the poor performance of the flash read API, we finally compromised to enable the XIP only in `hal_exflash_read()`

```
memcpy (XIP):
0000078766 [app] INFO: Read Time of 1MB data: 190 ms
0000079957 [app] INFO: Read Time of 1MB data: 190 ms
0000081147 [app] INFO: Read Time of 1MB data: 189 ms

FLASH_RxData:
0001106790 [app] INFO: Read Time of 1MB data: 909 ms
0001108700 [app] INFO: Read Time of 1MB data: 909 ms
0001110611 [app] INFO: Read Time of 1MB data: 909 ms
```

#### Crash issue

1. Compile and download prebootloader/bootloader/system-part1/user-app(flash_crusher)
2. On the Serial logging, you will notice `0000033706 [app] INFO: wrote: 1000 / 21000` where the user app is writing to fs
3. Send OTA using `particle flash <your_device_id> p2-tinker@5.1.0.bin` or any other large binary. 
4. Make sure the module won't get stuck in the OTA process

[flash_crusher.cpp.zip](https://github.com/particle-iot/device-os/files/9836066/flash_crusher.cpp.zip)
[p2-tinker@5.1.0.bin.zip](https://github.com/particle-iot/device-os/files/9836077/p2-tinker%405.1.0.bin.zip)

#### Cross-version upgrade

The modifications in `exflash_hal.cpp` shouldn't affect the prebootloader and bootloader.

Randomly upgrade/downgrade prebootloader/bootloader/system-part1 with the command `particle flash <your_device_id> <binary>`

### References

[CH110299]

---

### Completeness

- [x] User is totes amazing for contributing!
- [ ] Contributor has signed CLA ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md))
- [ ] Problem and Solution clearly stated
- [ ] Run unit/integration/application tests on device
- [ ] Added documentation
- [ ] Added to CHANGELOG.md after merging (add links to docs and issues)

### CODE REVIEW

File: hal/src/rtl872x/exflash_hal.cpp
Comment (avtolstoy) [inline hal/src/rtl872x/exflash_hal.cpp:44]: Let's keep all of this inside the class. If there is a static entry in the MPU table that we are modifying, just make it a static member variable. For initial disabling of XIP, just add a new method to initialize things etc and do it in hal_exflash_disable_xip

File: hal/src/rtl872x/exflash_hal.cpp
Comment (avtolstoy) [inline hal/src/rtl872x/exflash_hal.cpp:197]: Does this worsen the performance of reads? Is it absolutely required here?

**Review by avtolstoy — CHANGES_REQUESTED** [review]

File: hal/src/rtl872x/exflash_hal.cpp
Comment (YutingYou) [reply hal/src/rtl872x/exflash_hal.cpp]: Not really required, it is a little bit overcaution, and every time we call DCache API, we should better make sure that the buffer is 32-byte aligned I'll remove it.

File: hal/src/rtl872x/exflash_hal.cpp
Comment (YutingYou) [reply hal/src/rtl872x/exflash_hal.cpp]: Reworked

File: platform/MCU/rtl872x/src/flash_mal.c
Comment (XuGuohui) [inline platform/MCU/rtl872x/src/flash_mal.c:732]: return `SYSTEM_ERROR_NOT_SUPPORTED`

**Review by XuGuohui — CHANGES_REQUESTED** [review]

File: platform/MCU/rtl872x/src/flash_mal.c
Comment (YutingYou) [reply platform/MCU/rtl872x/src/flash_mal.c]: 👍 Fixed

File: platform/MCU/rtl872x/src/flash_mal.c
Comment (YutingYou) [reply platform/MCU/rtl872x/src/flash_mal.c]: Fixed

**Comment by YutingYou** [comment]
I notice an issue that it is not able to reconnect to the wifi once the wifi connection is disconnected. Haven't confirmed if it is brought by this PR, but it doesn't unblock the review.

**Review by avtolstoy — APPROVED** [review]

**Comment by avtolstoy** [comment]
@YutingYou Can you share an app that reproduces the wifi issue here please?

**Comment by YutingYou** [comment]
@avtolstoy use this app  [flash_crusher.cpp.zip](https://github.com/particle-iot/device-os/files/9836066/flash_crusher.cpp.zip)

Try to disconnect the wifi connection by e.g. going far away or putting P2 in a shielded box.

**Comment by YutingYou** [comment]
I couldn't reproduce this issue when I ran the test at home. Perhaps the router refuses the auto reconnection due to the poor signal (I’m using 5G wifi in the office)?

**Comment by YutingYou** [comment]
@keeramis If you got time, could you please help run an OTA test to see if the crash issue is gone? If it works for you, I'll merge it. Thanks!

**Comment by jmwilson** [comment]
I tried testing this (merged this branch into my local `develop` after also merging `trackerm-proto2-eval` in) and causing the device to go into safe mode. `particle serial inspect` seems to confused about what's on the device, only showing the bootloader module #2:
```
[ 14:16 ] james@james-wilson-macbook-pro:~/device-os/modules% particle serial inspect
Platform: 28 - Tracker M
Modules
  Bootloader module #2 - version 3, main location, 303104 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
```

**Comment by YutingYou** [comment]
@jmwilson This is an issue brought When fixing the last comment, have fixed, please try it again

**Comment by keeramis** [comment]
@YutingYou Thanks for the fixes. I tested on my form factor board with an OTA and I am not running into the issue anymore. I also ran this on my proto2 board with @jmwilson's custom proto2 branch + your PR commits, and it's fixed on proto2 boards as well. I went back and forth a couple of times between develop and your PR to ensure this PR fixes it. Great work and thanks! 🙌

**Comment by jmwilson** [comment]
@YutingYou tested and working for me as well!

**Comment by scott-brust** [comment]
I tested OTA of `prebootloader-part`, `bootloader` and `system-part1` from release `5.1.0` --> this branch (ie a build with the XIP changes) --> back to `5.1.0`. 
I bumped the prebootloader, bootloader and system versions by 1 each in order to distinguish the update. 
Here is `particle serial inspect` for 5.1.0:

```
Platform: 32 - P2
Modules
  Bootloader module #0 - version 2001, main location, 65536 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
      Bootloader module #2 - version 3
  Bootloader module #1 - version 2, main location, 8192 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
  Bootloader module #2 - version 3, main location, 303104 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
  System module #1 - version 5004, main location, 2097152 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
      Bootloader module #0 - version 2001
  User module #1 - version 6, main location, 2097152 bytes max size
    UUID: 439E4852F6005ABF343953CC8D02FDA7DDB08FF240F4386B9D6E7DF64E46B0A6
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
      System module #1 - version 5004
```

I did `particle flash <device_id> <binary>` for 1. prebootloader-part 2. bootloader and 3. system-part1. Each OTA'd in turn:
```
SYSTEM PART1 OTA:
Platform: 32 - P2
Modules
  Bootloader module #0 - version 2002, main location, 65536 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
      Bootloader module #2 - version 3
  Bootloader module #1 - version 2, main location, 8192 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
  Bootloader module #2 - version 4, main location, 303104 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
  System module #1 - version 5005, main location, 2097152 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
      Bootloader module #0 - version 2001
  User module #1 - version 6, main location, 2097152 bytes max size
    UUID: 439E4852F6005ABF343953CC8D02FDA7DDB08FF240F4386B9D6E7DF64E46B0A6
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
      System module #1 - version 5004
```
You can see each module has updated

After that I downgraded it back to 5.1.0 without any issue. Looks good to me 👍

**Comment by avtolstoy** [comment]
nice, thanks!

**Comment by YutingYou** [comment]
Thank you, folks, I also ran an overnight test, and didn't see the wifi disconnection in my home, I'll rebase and merge this PR today.

**Comment by XuGuohui** [comment]
Unfortunately my P2 eval board got disconnected from AP and it cannot reconnect over the whole night while keeping blinking green.

**Comment by scott-brust** [comment]
@YutingYou @XuGuohui Are there any concerns with merging this PR? 
Do you think the wifi disconnect problems are related to these changes? If they are not, can we address them in a separate PR?

**Comment by avtolstoy** [comment]
@YutingYou Please rebase and merge. We can tackle WiFi issues separately.

**Comment by YutingYou** [comment]
Sorry folks, I missed the message, I did two wifi re-connection tests on my two P2 modules on the PR multi-step priority disinheritance, one disabled the USB function can reproduce the re-connection issue after 12 hours, and the other one disabled the BLE function can keep connection after 60 hours.

So the BLE might be the issue, I’ll rebase and merge this PR first, and rerun the test as long as we merge the BLE PR.

cc: @scott-brust

**Review by XuGuohui — APPROVED** [review]

**Pull request merged**
